source/extensions/filters/http/custom_response/custom_response_filter.cc

```diff
@@ -0,0 +1,170 @@
+    if (!remote_data_source.has_value()) {
+      ENVOY_LOG(trace, "RemoteDataSource is empty");
+      config_->stats().custom_response_redirect_invalid_uri_.inc();
+      return Http::FilterHeadersStatus::Continue;
+    }
+    Http::Utility::Url absolute_url;
+    if (!absolute_url.initialize(remote_data_source->uri(), false)) {
+      ENVOY_LOG(trace, "Redirect for custom response failed: invalid location {}",
+                remote_data_source->uri());
+      config_->stats().custom_response_redirect_invalid_uri_.inc();
+      return Http::FilterHeadersStatus::Continue;
+    }
+    // Don't change the scheme from the original request
+    const bool scheme_is_http =
+        schemeIsHttp(*downstream_headers_, decoder_callbacks_->connection());
+
+    // Cache original host and path
+    const std::string original_host(downstream_headers_->getHostValue());
+    const std::string original_path(downstream_headers_->getPathValue());
+    const bool scheme_is_set = (downstream_headers_->Scheme() != nullptr);
+    Cleanup restore_original_headers([this, original_host, original_path, scheme_is_set,
+                                      scheme_is_http]() {
+      downstream_headers_->setHost(original_host);
+      downstream_headers_->setPath(original_path);
+      if (scheme_is_set) {
+        downstream_headers_->setScheme(scheme_is_http ? Http::Headers::get().SchemeValues.Http
+                                                      : Http::Headers::get().SchemeValues.Https);
+      }
+    });
+
+    // Replace the original host, scheme and path.
+    downstream_headers_->setScheme(absolute_url.scheme());
+    downstream_headers_->setHost(absolute_url.hostAndPort());
+
+    auto path_and_query = absolute_url.pathAndQueryParams();
+    if (Runtime::runtimeFeatureEnabled(
+            "envoy.reloadable_features.http_reject_path_with_fragment")) {
+      // Envoy treats internal redirect as a new request and will reject it if URI path
+      // contains #fragment. However the Location header is allowed to have #fragment in URI path.
+      // To prevent Envoy from rejecting internal redirect, strip the #fragment from Location URI if
+      // it is present.
+      auto fragment_pos = path_and_query.find('#');
+      path_and_query = path_and_query.substr(0, fragment_pos);
+    }
+    downstream_headers_->setPath(path_and_query);
+
+    if (decoder_callbacks_->downstreamCallbacks()) {
+      decoder_callbacks_->downstreamCallbacks()->clearRouteCache();
+    }
+    const auto route = decoder_callbacks_->route();
+    // Don't allow a redirect to a non existing route.
+    if (!route) {
+      config_->stats().custom_response_redirect_no_route_.inc();
+      ENVOY_LOG(trace, "Redirect for custom response failed: no route found");
+      return Http::FilterHeadersStatus::Continue;
+    }
+    downstream_headers_->setMethod(Http::Headers::get().MethodValues.Get);
+    downstream_headers_->remove(Http::Headers::get().ContentLength);
```

Comment:
I think the issue is that this will leak any headers added by the current filter chain to the redirected origin. This could be a problem for GFE3 or RXLB.

Reply:
Will run integration tests in coordination with GFE team.